lib/signs/utilities/early_am_suppression.ex

```diff
@@ -0,0 +1,308 @@
+          # - The headway top message will get filtered out and headways will be re-fetched
+          # - The prediction can now be filtered out or trigger the special case logic in
+          #   do_early_am_suppression since we reset the zone to "m" and the generic paging
+          #   will be reconstructed
+          {%Message.Headways.Paging{
+             destination: headway_top.destination,
+             range: headway_bottom.range
+           }, %{prediction | zone: "m"}}
+
+        _ ->
+          messages
+      end
+
+    {top_status, bottom_status} = statuses
+
+    Enum.map(
+      [
+        {top_message, top_scheduled, top_status},
+        {bottom_message, bottom_scheduled, bottom_status}
+      ],
+      fn {message, scheduled, status} ->
+        if(status == :none,
+          do: {message, %Message.Empty{}},
+          else: get_early_am_content(sign, {message}, scheduled, status, current_time)
+        )
+        |> case do
+          # If line has status :none, then it could be a paging headway message
+          {%Headways.Paging{destination: destination, range: range}, _} ->
+            {%Headways.Top{destination: destination, vehicle_type: :train},
+             %Headways.Bottom{range: range}}
+
+          content ->
+            content
+        end
+      end
+    )
+    |> List.to_tuple()
+  end
+
+  defp get_early_am_content(
+         sign,
+         messages,
+         {scheduled, destination},
+         status,
+         current_time
+       ) do
+    cond do
+      status == :fully_suppressed ->
+        {%EarlyAm.DestinationTrain{destination: destination},
+         %EarlyAm.ScheduledTime{
+           scheduled_time: scheduled
+         }}
+
+      status == :partially_suppressed ->
+        case filter_early_am_messages(messages, sign.id) do
+          [] ->
+            # If no valid predictions, try fetching headways
+            case Signs.Utilities.Headways.get_messages(sign, current_time) do
+              # If no headways are returned, default to timestamp message
+              {%Message.Empty{}, %Message.Empty{}} ->
+                {%EarlyAm.DestinationTrain{
+                   destination: destination
+                 },
+                 %EarlyAm.ScheduledTime{
+                   scheduled_time: scheduled
+                 }}
+
+              {headway_top, headway_bottom} ->
+                # Make sure routes is nil so that destination is used as headsign
+                {%{headway_top | destination: destination, routes: nil}, headway_bottom}
+            end
+
+          [message] ->
+            {message, %Content.Message.Empty{}}
+
+          messages ->
+            List.to_tuple(messages)
+        end
+    end
+  end
+
+  defp filter_early_am_messages(messages, sign_id) do
+    Tuple.to_list(messages)
+    |> Enum.reject(fn
+      message ->
+        match?(%Headways.Top{}, message) or match?(%Headways.Bottom{}, message) or
```

Comment:
The reason I did this is so that when one of the mezzanine source's directions is on headways but the other direction is on the timestamp message, we can easily put the two pages together using the `GenericPaging` struct a little further down the pipeline. Because we pass the top and bottom line into `get_early_am_content/5` separately, I felt that this approach was simpler than trying to stitch back together the top and bottom lines of the headway message if one of the lines happens to be using the timestamp message.